### ISSUE TITLE
Cannot parse AWS identity PKCS7 error retrieving from AWS [with IMDSv2 mandatory]

### ISSUE THREAD

**Issue by daz10000** [ncbi/sra-tools#829]
## TL;DR : 

Leaving the full story here for posterity, but the root cause is that AWS is using a newer standard for instance metadata `IMDSv2` that is session oriented, so the code fails to get the instance certificate.  It needs to use the session oriented api - get a token then get the certificate.   Workaround is to let instance support the old and new protocols via ec2 console but this is not ideal.

## Problem

I just installed 3.0.6 from source on an AWS instance running `aarch64` and  the installation successfully retrieves the 'hello world' test of two reads with `fastq-dump --stdout -X 2 SRR390728`.   I then added the aws credentials (via `vdb-config -i`  initially but also reviewed those in an editor and also tried setting the aws credentials via the `vdb-config` command line.  With all of these setups, any attempts to fetch the same example above (also trying to use prefetch on a different sample),  I get the following error message

```
2023-07-21T06:36:42 fastq-dump.3.0.6 err: error unexpected while resolving query within virtual file system module - Cannot parse AWS identity PKCS7 part, err=0 ( 500 )
Failed to call external services.
```

It is unclear why it's failing.  I tried looking with strace and the underlying problem seems to be a failed interaction with the host
(edit: this in retrospect was a red herring - see notes below)
```
Name:    sweb-vip.st-va.ncbi.nlm.nih.gov
Address:  165.112.7.16
```

After that happened,  I ran the tests for `ncbi-vdb` and 3 of them failed,  notably `Test_CLOUD_aws`

```
98% tests passed, 3 tests failed out of 130

Total Test time (real) = 216.71 sec

The following tests FAILED:
         25 - Test_KLIB_pack (Failed)
         44 - Test_CLOUD_aws (Failed)
        104 - Test_VFS_caching (Failed)
```

The output for test 44 shows this
```
44/130 Test: Test_CLOUD_aws
Command: "/home/ec2-user/ncbi-outdir/ncbi-vdb/linux/gcc/arm64/rel/obj/test-bin/Test_CLOUD_aws"
Directory: /home/ec2-user/src/ncbi-vdb/test/cloud
"Test_CLOUD_aws" start time: Jul 21 06:31 UTC
Output:
----------------------------------------------------------
Running 8 test cases...
TryPost (0 s)
TestBase64InIdentityDocument (0 s)
TestWrapInIdentityPkcs7 (0 s)
TestBase64InIdentityPkcs7 (0 s)
TestBase64MakeLocation (0 s)
/home/ec2-user/src/ncbi-vdb/test/cloud/test-aws.cpp:231: fatal error in "GetPkcs7": critical check 1000u < len failed [1000 >= 0]
GetPkcs7 (0 s)
PrintLocation (0 s)
CallCloudAddComputeEnvironmentTokenForSigner (0 s)

*** 1 failure detected in test suite "AwsTestSuite"
<end of output>
Test time =   1.03 sec
----------------------------------------------------------
Test Failed.
"Test_CLOUD_aws" end time: Jul 21 06:31 UTC
"Test_CLOUD_aws" time elapsed: 00:00:01
----------------------------------------------------------
```

That test is requiring the result of (I think the fetch to http://169.254.169.254/latest/dynamic/instance-identity/pkcs7 ) to return a string between 1000 and 3000 characters.  Running a curl, I don't get anything back - maybe the machine is misconfigured though it has general network connectivity.

I reran the test with some instrumentation, and the returned pkcs7 information is indeed empty.

Running a curl against that ip, I get a 401

```
curl --verbose http://169.254.169.254
*   Trying 169.254.169.254:80...
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET / HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/8.0.1
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< Content-Length: 0
< Date: Fri, 21 Jul 2023 06:50:54 GMT
< Server: EC2ws
< Connection: close
< Content-Type: text/plain
<
* Closing connection 0
```

So reading a little more, that looks like it's an attempt to get the instance metadata, and I need to find out why that's failing.  If you have any tips, that would be much appreciated.  I saw a similar report #526 a while back with no resolution

**Comment by daz10000** [comment]
Digging a little more into the AWS instance meta-data apis,  their examples all fetch a token first, and then supply the token to the pkcs7 call,  for example using CURL

```
 TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"` && curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/dynamic/instance-identity/pkcs7
```

This successfully retrieves the PKCS7 data on my machine, but simply trying to go straight to the pkcs7 endpoint fails with a 401.  Is it possible that the code needs an update to accommodate AWS security changes? or have I misconfigured something.

**Comment by daz10000** [comment]
And yes... confirmed.  If I modify the instance to make `IMDsV2` optional, instead of mandatory,  fetching resumes.  I can do that for now, but ideally the code would support the newer metadata standard gracefully.  Thanks for listening to me rant..

**Comment by klymenko** [comment]
Thank you for reporting.
This issue was fixed in our [release 3.0.7](https://github.com/ncbi/sra-tools/wiki/01.-Downloading-SRA-Toolkit).
Please confirm.

**Comment by daz10000** [comment]
sorry for delay replying,  but I was waiting for another opportunity to download.  It  appears to be still broken. Here is what I did to test

1. changed metadata  preference on the ec2  instance back to make version 2 mandatory rather than optional

![image](https://github.com/ncbi/sra-tools/assets/6342645/44ff9353-1f74-46a2-abbb-023e8e781f14)

2. git pull on `sra-tools`  and `ncbi-vdb` repos
3. make and then make install for `ncbi-vdb` and `sra-tools`
4. Tried another prefetch

 the output looks the same as the previous problem

`2023-09-10T17:41:00 prefetch.3.0.7 err: error unexpected while resolving query within virtual file system module - Cannot parse AWS identity PKCS7 part, err=0 ( 500 )`

Prefetch is reporting version 3.0.7

```
prefetch --version

prefetch : 3.0.7
```

I didn't do a full build from scratch, and could try that but I'm fairly confident it's not fixed.  If there is any code or test I can run to verify I'm running the new version, let me know.  for now I'll disable v2 metadata on the ec2 as a workaround  (and that does still work as before).

**Comment by klymenko** [comment]
Could you download and try our build?

**Comment by daz10000** [comment]
Again thanks for your patience,  I had a chance to test again this evening.  I used the yum setup approach from the page you link.  Unfortunately it installed an x64 binary but I'm running an ARM aarch64 operating system.  Looking at the ftp site, it doesn't look like you have a prebuilt binary for that platform.

https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/3.0.7/setup-yum.sh

![image](https://github.com/ncbi/sra-tools/assets/6342645/f4c63770-0029-4928-b41a-02d085b07297)

**Comment by klymenko** [comment]
What is the output of `prefetch -V` ?

**Comment by daz10000** [comment]
```
[ec2-user@ip-10-0-19-24 src]$ prefetch -V

prefetch : 3.0.7
```

**Comment by daz10000** [comment]
Would you be able to point me at the piece of the code that implements the IMDSv2 protocol quickly - I want to sanity check that the code is in my current source tree.  I'll take a quick look now but pointers appreciated.  THis is starting to become a headache - for a single ec2 machine I can downgrade IMDS but it's harder on a cluster and not great from a security standpoint, so I'd like to get to V2 asap

**Comment by daz10000** [comment]
And for good measure, I just did a git pull on the sra-tools and ncbi-vdb repos, built and installed, and so I'm at 3.0.8

```
[ec2-user@ip-10-0-19-24 tmp]$ prefetch -V

prefetch : 3.0.8
```

Same problem

```
[ec2-user@ip-10-0-19-24 tmp]$ fasterq-dump SRR000001
2023-09-19T20:02:27 fasterq-dump.3.0.8 err: error unexpected while resolving query within virtual file system module - Cannot parse AWS identity PKCS7 part, err=0 ( 500 )
Failed to call external services.
```

**Comment by klymenko** [comment]
What is the output of `srapath SRR000001` ?

**Comment by klymenko** [comment]
[GetInstanceInfo](https://github.com/ncbi/ncbi-vdb/blob/master/libs/cloud/aws.c#L114) implements reading AWS identity token.

**Comment by daz10000** [comment]
As a sanity check,  I reran the ncbi-vdb test suite as well.  In the original report above, 3 tests failed.  Looks like this one is fixed `104 - Test_VFS_caching (Failed)`, 

Now we have just two tests failing.  Test 45 in particular is exercising the IMDS functionality I think, although if it's just checking if V1 works, it will of course fail 

```
The following tests FAILED:
         25 - Test_KLIB_pack (Failed)
         45 - Test_CLOUD_aws (Failed)
```

## Test 45

```
45/134 Test: Test_CLOUD_aws
Command: "/home/ec2-user/ncbi-outdir/ncbi-vdb/linux/gcc/arm64/rel/obj/test-bin/Test_CLOUD_aws"
Directory: /home/ec2-user/src/ncbi-vdb/test/cloud
"Test_CLOUD_aws" start time: Sep 19 20:08 UTC
Output:
----------------------------------------------------------
Running 10 test cases...
TestBase64InIdentityDocument (0 s)
TestWrapInIdentityPkcs7 (0 s)
TestBase64InIdentityPkcs7 (0 s)
TestBase64MakeLocation (0 s)
***IMDSv=1
Get_IMDS_version (0 s)
/home/ec2-user/src/ncbi-vdb/test/cloud/test-aws.cpp:242: fatal error in "Get_Pkcs7": critical check 1000u < len failed [1000 >= 0]
Get_Pkcs7 (0 s)
PrintInstance_NotAllowed (0 s)
***ce_token=LS0tLS1CRUdJTiBQS0NTNy0tLS0tCgotLS0tLUVORCBQS0NTNy0tLS0tCg==.
PrintInstance_Allowed (0 s)
CallCloudAddComputeEnvironmentTokenForSigner (1 s)
***location=s3.
GetLocation (0 s)

*** 1 failure detected in test suite "AwsTestSuite"
<end of output>
Test time =   1.09 sec
----------------------------------------------------------
Test Failed.
"Test_CLOUD_aws" end time: Sep 19 20:08 UTC
"Test_CLOUD_aws" time elapsed: 00:00:01
----------------------------------------------------------
```


## 25

```
25/134 Test: Test_KLIB_pack
Command: "/home/ec2-user/ncbi-outdir/ncbi-vdb/linux/gcc/arm64/rel/obj/test-bin/Test_KLIB_pack"
Directory: /home/ec2-user/src/ncbi-vdb/test/klib
"Test_KLIB_pack" start time: Sep 19 20:08 UTC
Output:
----------------------------------------------------------
Running 28 test cases...
Pack_null_psize (0 s)
Pack_bad_unpacked (0 s)
Pack_bad_packed (0 s)
Pack_0_ssize (0 s)
Pack_bad_ssize (0 s)
Pack_small_dsize_no_consumed (0 s)
Pack_small_dsize (0 s)
Pack_null_dest (0 s)
Pack_memcpy (0 s)
Unpack_null_usize (0 s)
Unpack_bad_unpacked (0 s)
Unpack_ssize_0 (0 s)
Unpack_ssize_low (0 s)
Unpack_null_src (0 s)
Unack_small_dsize_no_consumed (0 s)
Unack_null_usize (0 s)
Pack_Unpack_memcpy (0 s)
Pack_8_Unpack_single (0 s)
Pack_8_Unpack_multiple (0 s)
Pack_16_Unpack_single (0 s)
Pack_16_Unpack_multiple (0 s)
Pack_32_Unpack_single (0 s)
Pack_32_Unpack_multiple (0 s)
Pack_64a_Unpack_single (0 s)
Pack_64b_Unpack_single (0 s)
Pack_64a_Unpack_multiple (0 s)
/home/ec2-user/src/ncbi-vdb/test/klib/test-pack.cpp:335: fatal error in "Pack_64b_Unpack_multiple": critical check src == (PackUnpackMultiple( src, packed )) failed
Pack_64b_Unpack_multiple (0 s)
Pack_fail (0 s)

*** 1 failure detected in test suite "KlibPackTestSuite"
<end of output>
```

**Comment by klymenko** [comment]
What OS are you running?

**Comment by daz10000** [comment]
`Linux ip-10-0-19-24.us-west-2.compute.internal 6.1.34-59.116.amzn2023.aarch64 #1 SMP Thu Jun 29 18:11:49 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux`
`

**Comment by daz10000** [comment]
> srapath SRR000001

I noticed I missed this above.  It also fails on the metadata check

```
[ec2-user@ip-10-0-19-24 ncbi-vdb]$ srapath SRR000001
2023-09-19T21:16:55 srapath.3.0.8 err: error unexpected while resolving query within virtual file system module - Cannot parse AWS identity PKCS7 part, err=0 ( 500 )
```

**Comment by daz10000** [comment]
I'd like to start instrumenting this test which basically reproducing the failure.   Despite reading for 30 mins,  I still have no idea how to run just one test with the cmake test setup.  Is there an easy incantation to run just that failing test 45?

![image](https://github.com/ncbi/sra-tools/assets/6342645/236774e3-19c4-4087-a894-052b3df3d79c)

**Comment by klymenko** [comment]
I was able to reproduce it.
There is a difference between x86 and arm.
It works on x86.

**Comment by klymenko** [comment]
I will give you a quick fix tomorrow.

**Comment by daz10000** [comment]
Do you know the problem? - I started instrumenting the calls and the first failure is that the getinstanceinfo call seems to go into case 1 here which is problematic

![image](https://github.com/ncbi/sra-tools/assets/6342645/777cfaea-c129-4db1-89e7-59a0c0dfe013)

**Comment by klymenko** [comment]
Here is the fix:
```
[ec2-user@ip-172-31-26-70 ncbi-vdb]$ git diff
diff --git a/libs/cloud/aws.c b/libs/cloud/aws.c
index 3584a735..a315f166 100644
--- a/libs/cloud/aws.c
+++ b/libs/cloud/aws.c
@@ -82,7 +82,8 @@ IMDS_version( const KNSManager * kns, char *buffer, size_t bsize )
     // try IMDSv1 first
     if ( KNSManager_Read( kns, buffer, bsize,
                           INSTANCE_URL_PREFIX "meta-data",  HttpMethod_Get,
-                          NULL, NULL) == 0 )
+                          NULL, NULL) == 0 
+                   && buffer[0] != '\0' )
     {
         buffer[ 0 ] = 0;
         return 1;
```

**Comment by daz10000** [comment]
And that's a confirmed kill... thanks! and I appreciate your persistence on this one.

**Comment by daz10000** [comment]
Thanks again for the diligence and persistence.  Closing this to tidy your ticket queue but please adjust as you see fit.  I have working binary compiled from git with your patch manually applied, so I consider it solved, but happy to test future code if helpful.

**Comment by klymenko** [comment]
Thank you for finding it.
We will include this fix into our next release 3.0.9.
